rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,660 @@
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  epoch int [not null]
+  slot int [not null]
+  ...
+  columns
+  ledger_proof_nonce int [not null]
+  status consensus_status
+  block_length int [not null]
+}
+
+Table user_commands {
+  id string [pk]
+  is_delegation bool [not null]
+  nonce int64 [not null]
+  from string [not null]
+  to string [not null]
+  amount int64 [not null]
+  fee int64 [not null]
+  memo string [not null]
+  time_first_seen date
+  is_manually_added bool [not null]
+}
+
+Table fee_transfers {
+  id string [pk]
+  fee int64 [not null]
+  receiver string [not null]
+  time_first_seen date
+  is_manually_added bool [not null]
+}
+
+Table block_to_transactions {
+  state_hash string [ref: > block.state_hash]
+  transaction_id int [not null, ref: > user_commands.id, ref: > fee_transfers.id]
+  time_seen_transaction date
+  sender string
+  receiver string [not null]
+  block_length length
+    Indexes {
+    state_hash [name:"block"]
+    transaction_id [name: "transaction"]
+    (receiver, block_length, epoch, slot) [name: "fast_receiver_block_compare_pagination"]
+    (sender, block_length, epoch, slot) [name: "fast_sender_block_compare_pagination"]
+    (receiver, time_seen_transaction) [name: "fast_receiver_time_seen_transaction_pagination"]
+    (sender, time_seen_transaction) [name: "fast_sender_time_seen_transaction_pagination"]
+  }
+  // Rows are uniquely identified by `state_hash` and `transaction_id`
+}
+
+
+// Use Rocksdb for this
+Table receipt_chain_hashes {
+  hash string [pk]
+  previous_hash string
+}
+```
+
+Below is an image of the relationships of the schemas:
+
+![Client process](../docs/res/coda_sqlite.png)
+
+Here is a link of an interactive version of the schema: https://dbdiagram.io/d/5d30b14cced98361d6dccbc8
+
+Notice that all the possible joins that we have to run through the `block_to_transactions` table. The `transaction_id` and `state_hash` columns are indexed to make it fast to compute which transactions are in an block and which blocks have a certain transaction for the `transactionStatus` query. We have multicolumn indexes for `(receiver, block_length, epoch, slot)` and `(sender, block_length, epoch, slot)` to boost the performance of pagination queries where are sorting based on `block_compare`.
+
+We also have multicolumn indexes for `(receiver, time_seen_transaction)` and `(sender, time_seen_transaction)` to boost the performance of pagination queries where we are sorting based on the time we first see a transaction. The join table has quite a bit of indexes, which will impact the performance of the table when performing write updates. If the writes are too slow, then the join table should have only one type of pagination comparison. We should prefer `block_compare` over `time_seen_transaction` since `time_seen_transaction` is a mutable field and the change to this field would cause more index updates to `time_seen_transaction`.
```

Comment:
How slow do you think this will be? I can't imagine it would be too slow even with all possible indices

Reply:
I am not sure. We would have to experiment with this.